src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+                    }
+
+                    CheckpointInfo checkpoint = this.checkpoints.GetCheckpoint(currentChainedHeader.Height);
+                    if (checkpoint != null)
+                    {
+                        this.logger.LogDebug("Chained header '{0}' is a checkpoint.", currentChainedHeader);
+
+                        connectedHeaders = this.HandleCheckpointsHeader(currentChainedHeader, latestNewHeader, checkpoint);
+                        break;
+                    }
+
+                    currentChainedHeader = currentChainedHeader.Previous;
+                }
+
+                if ((connectedHeaders == null) && isWithinCheckpoints)
+                {
+                    connectedHeaders = new ConnectedHeaders() {Consumed = latestNewHeader};
+                    this.logger.LogTrace("Chained header '{0}' bellow last checkpoint.", currentChainedHeader);
+                }
+
+                if (connectedHeaders != null)
+                {
+                    this.logger.LogTrace("(-)[CHECKPOINT_OR_ASSUMED_VALID]:{0}", connectedHeaders);
+                    return connectedHeaders;
+                }
+            }
+
+            if (latestNewHeader.ChainWork > this.chainState.ConsensusTip.ChainWork)
+            {
+                this.logger.LogDebug("Chained header '{0}' is the tip of a chain with more work than our current consensus tip.", latestNewHeader);
+
+                connectedHeaders = this.MarkBetterChainAsRequired(latestNewHeader);
+            }
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// A chain with more work than our current consensus tip was found so mark all it's descendants as required.
+        /// </summary>
+        /// <param name="latestNewHeader">The new header that represents a longer chain.</param>
+        /// <returns>The new headers that need to be downloaded.</returns>
+        private ConnectedHeaders MarkBetterChainAsRequired(ChainedHeader latestNewHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(latestNewHeader), latestNewHeader);
+
+            ConnectedHeaders connectedHeaders = new ConnectedHeaders();
+            connectedHeaders.DownloadTo = connectedHeaders.Consumed = latestNewHeader;
+
+            ChainedHeader current = latestNewHeader;
+            ChainedHeader next = current;
+
+            while (this.HeaderWasNotRequested(current))
+            {
+                current.BlockDataAvailability = BlockDataAvailabilityState.BlockRequired;
+
+                next = current;
+                current = current.Previous;
+            }
+
+            connectedHeaders.DownloadFrom = next;
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// Mark all previous blocks to <see cref="chainedHeader"/> as <see cref="ValidationState.AssumedValid"/>.
```

Comment:
to -> of

see -> paramref

Reply:
And it should also include the final header itself, so it should say something like Mark the chain ending with...